source/auth/auth.rst

```diff
@@ -1159,6 +1159,205 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+Note that the principal name is optional as it may be provided by the IDP in environments where only one IDP is used.
+
+.. code:: idl
+
+  OIDCMechanismServerStep1:
+    description: "Server's reply to clientStep1"
+    strict: false
+    fields:
+      authorizeEndpoint:
+        description: >-
+          URL where the IDP may be contacted for end user
+          authentication and authorization code generation.
+        type: string
+        optional: true # Req if deviceAuthorizeEndpoint not present
+      tokenEndpoint:
+        description: >-
+          URL where the IDP may be contacted for authorization
+          code <=> ID/access token exchange.
+        type: string
+        optional: true # Req if deviceAuthorizeEndpoint not present
+      deviceAuthorizeEndpoint:
+        description: >-
+          URL where the IDP may be contacted for device
+          authentication and authorization code generation.
+        type: string
+        optional: true # Req if authorizeEndpoint not present
+      clientId:
+        description: "Unique client ID for this OIDC client"
+        type: string
+      clientSecret:
+        description: "Secret used when communicating with IDP"
+        type: string
+        optional: true
+      requestScopes:
+        description: "Additional scopes to request from IDP"
+        type: array<string>
+        optional: true
+
+Server will use principalName (n) if provided in clientStep1 to select an appropriate IDP.  This IDP's configuration will be returned in serverStep1 to instruct the client on how to acquire an Access Token.  This Access Token will be used in clientStep2 to complete authentication.  The principalName identified within the Access Token MUST match any provided during clientStep1, or the authentication operation will fail.
+
+.. code:: idl
+
+  OIDCMechanismClientStep2:
+      description: "Client's request with signed token"
+      strict: false
+      fields:
+          jwt:
+              description: "Compact serialized JWT with signature"
+              cpp_name: JWT
+              type: string
+
+The driver MAY, if it has previously acquired a valid Access Token, or has cached the reply from serverStep1, begin their authentication by sending clientStep2 with their initial saslStart request and authenticate in a single round trip.  Similarly, as an authenticated client's token's expirationTime approaches, it should proactively acquire a new Access Token, and it MAY do so with cached IDP details, skipping directly to clientStep2.
+
+
+`MongoCredential`_ Properties
+`````````````````````````````
+
+username
+    MUST NOT be specified.
+
+source
+    MUST be "$external". Defaults to ``$external``.
+
+password
+    MUST NOT be specified.
+
+mechanism
+    MUST be "MONGODB-OIDC"
+
+mechanism_properties
+    PRINCIPAL_NAME
+        Drivers MUST allow the user to specify a principal name, which
+        is required by the server when more than one Identity Provider
+        is configured.
```

Comment:
It’s a surprise that we expect `PRINCIPAL_NAME` as an auth mechanism property instead of passing it in the username. Is there a specific reason for this? As I understand it, it fulfills the same role as the username, and in particular also matches the role of the username for similar auth mechanisms like Kerberos or LDAP.

Reply:
I thought it might be misleading to use the `username` field, since the server is actually creating a derived username for this session.

Reply:
I guess the question then is about the semantics of the `username` field, whether it is “This is the identity I am attempting to authenticate as” or “This is the username string that the database uses and reports”. I would maybe still lean towards interpreting it as the former, although I see your point about this differing from the DB username.

Reply:
@sgolemon do you have any feelings on this? I'm vaguely reminded of how during LDAP authentication we map the presented authentication name into a name which the LDAP server will recognize through the UserToDNMapping.

I'm also leaning heavily toward the username in `mongodb://[username:password@]host1` is "This is the identity I am attempting to authenticate as" as Anna said

Reply:
I do have feelings! My assumption was that the principal name hint would and should be the `username`, and not a custom property as that seems like "least surprise" from an end-developer point of view.  It's true that the derived principal name may differ (depending on IDP and MongoDB configurations), but from the application point of view the synthesized principal name is an implementation detail and all we get by placing the hint into mechanism properties and prohibiting the username field is a little confusion about why it's not in the usual spot.

Reply:
From my understanding of the server technical design and implementation, the principal name is by default the "sub" field of the jwt token, which if the IdP uses pairwise identifier is obfuscated by a [salt](https://openid.net/specs/openid-connect-core-1_0.html#PairwiseAlg).

Reply:
So in that case it feels even less like a user name.

Reply:
@blink1073 On today's initiative sync we briefly touched on this, and the sentiment was to use the username field like we would for LDAP and Kerberos. Unless there are concerns with it that haven't been brought up so far, this shouldn't be an auth mechanism property.

Reply:
Okay fair, I will update the spec and connection string tests.

Reply:
Done